lambda/transl_array_comprehension.ml

```diff
@@ -0,0 +1,932 @@
+  let let_if_reused (type u) ~(usage : u t) let_kind value_kind name value
+      : lambda * (u, Let_binding.t) if_reused =
+    match usage with
+    | Once ->
+        value, Used_once
+    | Many ->
+        let var, binding =
+          Let_binding.make_var let_kind value_kind name value
+        in
+        var, Reusable binding
+end
+
+module Precompute_array_size : sig
+  (** Generates the lambda expression that throws the exception once we've
+      determined that precomputing the array size has overflowed.  The check for
+      overflow is done elsewhere; this just throws the exception
+      unconditionally. *)
+  val raise_overflow_exn : loc:scoped_location -> lambda
+
+  (** [safe_product_pos_vals ~loc xs] generates the lambda expression that
+      computes the product of all the lambda terms in [xs] assuming they are all
+      strictly positive (nonzero!) integers, failing if any product overflows
+      (equivalently, if the whole product would overflow).  This function must
+      look at its inputs multiple times, as they are evaluated more than once
+      due to the overflow check; the optional argument [variable_name]
+      customizes the string used to name these variables. *)
+  val safe_product_pos :
+    ?variable_name:string -> loc:scoped_location -> lambda list -> lambda
+end = struct
+  let raise_overflow_exn ~loc =
+    (* CR aspectorzabusky: Is this idiomatic?  Should the argument to [string]
+       (a string constant) just get [Location.none] instead? *)
+    let loc' = Debuginfo.Scoped_location.to_location loc in
+    let slot =
+      transl_extension_path
+        loc
+        Env.initial_safe_string
+        Predef.path_invalid_argument
+    in
+    (* CR aspectorzabusky: Should I call [Translprim.event_after] here?
+       [Translcore.asssert_failed] does (via a local intermediary). *)
+    Lprim(Praise Raise_regular,
+          [Lprim(Pmakeblock(0, Immutable, None, alloc_heap),
+                 [ slot
+                 ; string
+                     ~loc:loc'
+                     "integer overflow when precomputing the size of an array \
+                      comprehension" ],
+                 loc)],
+          loc)
+
+  (** [safe_mul_pos_vals ~loc x y] generates the lambda expression that computes
+      the product [x * y] of two strictly positive (nonzero!) integers and fails
+      if this overflowed; the inputs are required to be values, as they are
+      evaluated more than once *)
+  let safe_mul_pos_vals ~loc x y =
+    let open (val Lambda_utils.int_ops ~loc) in
+    let product, product_binding =
+      Let_binding.make_var (Immutable Alias) Pintval "product" (x * y)
+    in
+    (* [x * y] is safe, for strictly positive [x] and [y], iff you can undo the
+       multiplication: [(x * y)/y = x].  We assume the inputs are values, so we
+       don't have to bind them first to avoid extra computation. *)
+    Let_binding.let_one product_binding
+      (Lifthenelse(product / y = x,
```

Comment:
There's really no better way.  Not unless we get intrinsics for checking overflow.  But yeah, as far as I can tell this is the only way to test for multiplication overflow, and you're right to be unhappy.

Reply:
I'm not sure anything else can be done without extra support further through the compiler, which probably isn't worth it in the initial instance?

Reply:
There are ways to check for overflow in multiplication without a division (for positive integers).
One possible solution is to split the numbers in two (high and low bits), then if both high bits are 0 then there's no overflow, if both are non-zero there's definitively an overflow, and in the last case there's a bit of extra work to do but the checks are either half-width multiplication overflow checks (a simple comparison) or addition overflow checks.
This has the advantage of being quite cheap in the common case where none of the numbers are above 2^31.
If anyone is interested I can try to provide an actual implementation.

Reply:
While that sounds very cool, @lthls, I'm not sure it's worth implementing here.

A little context (that got lost in my initial volley of comments): I chatted with @gretay-js about the possibility of adding primitives to check for overflow, using flags within the CPU. She said that it might be possible. I then asked around the office if folks thought the feature was worthwhile, and the answer was "meh". That is, if we had this support, we'd use it, but it may not be worth the time necessary to design, implement, publicize, and maintain. However, I had already made the comments in this PR, and forgot to delete them. So, the sentiment remains -- gee, overflow checks are ugly -- but we've also grown uninterested in actually solving the problem. (@antalsz had this context when the comments arrived.)